internal/controller/gateway/gateway_controller.go

```diff
@@ -145,26 +159,188 @@ func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
+
+	// In order to preserve existing status conditions, we need copy an existing listener status
+	// if there is one.
+	listenerStatusByName := make(map[gatewayv1.SectionName]gatewayv1.ListenerStatus)
+	for _, l := range gw.Status.Listeners {
+		listenerStatusByName[l.Name] = l
+	}
+
+	// Copy the existing addresses and conditions from the existing status
+	newStatus := gatewayv1.GatewayStatus{
+		Addresses:  gw.Status.Addresses,
+		Conditions: gw.Status.Conditions,
+		Listeners:  []gatewayv1.ListenerStatus{},
+	}
+
+	// Listener Rules specific to the ngrok operator & ngrok product
+	// (1) ngrok does not support listening for HTTP/HTTPS traffic on arbitrary ports.
+	// Listeners that specify protocolTypeHTTP must listen on port 80.
+	// Listeners that specify protocolTypeHTTPS must listen on port 443.
+	// (2) ngrok does not support HTTP/HTTPS listeners without a hostname.
+	for _, l := range gw.Spec.Listeners {
+		listenerStatus, ok := listenerStatusByName[l.Name]
+		if !ok {
+			listenerStatus = gatewayv1.ListenerStatus{
+				Name:           l.Name,
+				SupportedKinds: []gatewayv1.RouteGroupKind{},
+				AttachedRoutes: 0,
+				Conditions:     []metav1.Condition{},
+			}
+		}
+
+		listenerValid := false
+
+		switch l.Protocol {
+		case gatewayv1.HTTPProtocolType:
+			switch {
+			case l.Port != 80:
+				meta.SetStatusCondition(&listenerStatus.Conditions, r.newListenerCondition(
+					gw,
+					gatewayv1.ListenerConditionAccepted,
+					gatewayv1.ListenerReasonPortUnavailable,
+					"ngrok only supports HTTP on port 80",
+				))
+			case l.Hostname == nil || *l.Hostname == "":
+				meta.SetStatusCondition(&listenerStatus.Conditions, r.newListenerCondition(
+					gw,
+					gatewayv1.ListenerConditionAccepted,
+					ListenerReasonHostnameRequired,
+					"ngrok does not support HTTP listeners without a hostname",
+				))
+			default:
+				listenerValid = true
+			}
+		case gatewayv1.HTTPSProtocolType:
+			switch {
+
+			case l.Port != 443:
+				meta.SetStatusCondition(&listenerStatus.Conditions, r.newListenerCondition(
+					gw,
+					gatewayv1.ListenerConditionAccepted,
+					gatewayv1.ListenerReasonPortUnavailable,
+					"ngrok only supports HTTPS on port 443",
+				))
+			case l.Hostname == nil || *l.Hostname == "":
+				meta.SetStatusCondition(&listenerStatus.Conditions, r.newListenerCondition(
+					gw,
+					gatewayv1.ListenerConditionAccepted,
+					ListenerReasonHostnameRequired,
+					"ngrok does not support HTTPS listeners without a hostname",
+				))
+			default:
+				listenerValid = true
+			}
+		default:
+			listenerValid = true
+		}
+
+		if listenerValid {
+			meta.SetStatusCondition(&listenerStatus.Conditions, r.newListenerCondition(
+				gw,
+				gatewayv1.ListenerConditionAccepted,
+				gatewayv1.ListenerReasonAccepted,
+				"",
+			))
+			meta.SetStatusCondition(&listenerStatus.Conditions, r.newListenerCondition(
+				gw,
+				gatewayv1.ListenerConditionProgrammed,
+				gatewayv1.ListenerReasonPending,
+				"",
```

Comment:
nit: is there a helpful message we could add here?

Reply:
Message has been added